packages/sdk/src/types/util.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * Discriminator for {@linkcode TypedJson}. The member used to identify each type is
+ * exported from each type module and can be accessed through named export `JSON_DISCRIMINATOR`.
+ */
+export enum TypedJsonDiscriminator {
+    AccountAddress = 'ccd_account_address',
+    BlockHash = 'ccd_block_hash',
+    CcdAmount = 'ccd_ccd_amount',
+    ContractAddress = 'ccd_contract_address',
+    ContractName = 'ccd_contract_name',
+    CredentialRegistrationId = 'ccd_cred_reg_id',
+    DataBlob = 'ccd_data_blob',
+    Duration = 'ccd_duration',
+    Energy = 'ccd_energy',
+    EntrypointName = 'ccd_entrypoint_name',
+    InitName = 'ccd_init_name',
+    ModuleReference = 'ccd_module_reference',
+    Parameter = 'ccd_parameter',
+    ReceiveName = 'ccd_receive_name',
+    ReturnValue = 'ccd_return_value',
+    SequenceNumber = 'ccd_sequence_number',
+    Timestamp = 'ccd_timestamp',
+    TransactionExpiry = 'ccd_transaction_expiry',
+    TransactionHash = 'ccd_transaction_hash',
+}
+
+/**
+ * Type describing the JSON representation of strong types used in the SDK.
+ *
+ * @template V - The serializable JSON value
+ */
+export interface TypedJson<V> {
+    /** The type discriminator */
+    ['@type']: TypedJsonDiscriminator;
+    /** The serializable type value */
+    value: V;
+}
+
+/**
+ * Base class for concordium domain types
+ *
+ * @template D - The JSON discriminator
+ * @template V - The serializable JSON value
+ */
+export abstract class TypeBase<V> implements ToTypedJson<V> {
+    protected abstract typedJsonType: TypedJsonDiscriminator;
+    protected abstract get serializable(): V;
+
+    public toTypedJSON(): TypedJson<V> {
+        return {
+            ['@type']: this.typedJsonType,
+            value: this.serializable,
+        };
+    }
+}
+
+/**
+ * Common interface implemented by strong types used in the SDK
+ * for converting instances of the type to types of {@linkcode TypedJson}.
+ *
+ * @template D - The JSON discriminator
+ * @template V - The serializable JSON value
+ */
+export interface ToTypedJson<V> {
+    /**
+     * Converts type to {@linkcode TypedJson}
+     *
+     * @returns {TypedJson} The typed JSON.
+     */
+    toTypedJSON(): TypedJson<V>;
```

Comment:
I ran into an issue where these types would be stringified before being sent to WASM, so that's the main reason. Also if users were to stringify types as part of f.x. http requests, I think it would be nice if they didn't have to deal with the discriminator stuff added here.

Will add a comment.